### PR TITLE
Feature/Cross validation

### PR DESCRIPTION
**Pull request by jduerholt**
This PR implements the possibility to cross validate models in bofire and then analyze the results. Plotting functionality will be added later. 

In addition a tutorial notebook to demonstrate the functionality is included.

### CODE REVIEW

File: bofire/models/diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+metrics = {
+    RegressionMetricsEnum.MAE: mean_absolute_error,
+    RegressionMetricsEnum.MSD: mean_squared_error,
+    RegressionMetricsEnum.R2: r2_score,
+    RegressionMetricsEnum.MAPE: mean_absolute_percentage_error,
+}
+
+
+class CVResult(PydanticBaseModel):
```
Comment (bertiqwerty) [inline bofire/models/diagnostics.py:24]: Calling this `CvResult` increases readability at least for me. See what Google proposes  https://google.github.io/styleguide/jsguide.html#naming-camel-case-defined.

File: bofire/models/diagnostics.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        return v
+
+    @validator("uncertainty")
+    def validate_uncertainty(cls, v, values):
+        if not is_numeric(v):
+            raise ValueError("Not all values of observed are numerical")
+        return v
+
+    @property
+    def num_samples(self) -> int:
```
Comment (bertiqwerty) [inline bofire/models/diagnostics.py:72]: we should maybe be consistent throughout bofire. Either always `num_something` or `n_something` for number of something. I prefer the shorter one, of course ;).

File: bofire/models/diagnostics.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+        Returns:
+            float: Metric value.
+        """
+        if self.num_samples == 1:
+            raise ValueError("Metric cannot be calculated for only one sample.")
+        return metrics[metric](self.observed.values, self.predicted.values)
+
+
+class CVResults(PydanticBaseModel):
```
Comment (bertiqwerty) [inline bofire/models/diagnostics.py:94]: As above.

**Review by bertiqwerty — APPROVED** [review]
Thanks. Approved. Just left some comments.

File: bofire/models/diagnostics.py (hunk 3 above)
Comment (jduerholt) [reply bofire/models/diagnostics.py]: Done.

**Pull request merged**